activitysim/benchmarking/componentwise.py

```diff
@@ -0,0 +1,601 @@
+    All this happens here, before the model component itself
+    is actually executed inside the timed portion of the loop.
+    """
+    if isinstance(configs_dirs, str):
+        configs_dirs = [configs_dirs]
+    inject.add_injectable(
+        "configs_dir", [os.path.join(working_dir, i) for i in configs_dirs]
+    )
+    inject.add_injectable("data_dir", os.path.join(working_dir, data_dir))
+    inject.add_injectable("output_dir", os.path.join(working_dir, output_dir))
+
+    reload_settings(
+        settings_filename,
+        benchmarking=component_name,
+        checkpoints=False,
+    )
+
+    component_logging(component_name)
+    logger.info("connected to component logger")
+    config.filter_warnings()
+    logging.captureWarnings(capture=True)
+
+    # register abm steps and other abm-specific injectables outside of
+    # benchmark timing loop
+    if not inject.is_injectable("preload_injectables"):
+        logger.info("preload_injectables yes import")
+        from activitysim import abm
+    else:
+        logger.info("preload_injectables no import")
+
+    # Extract the resume_after argument based on the model immediately
+    # prior to the component being benchmarked.
+    models = config.setting("models")
+    try:
+        component_index = models.index(component_name)
+    except ValueError:
+        # the last component to be benchmarked isn't included in the
+        # pre-checkpointed model list, we just resume from the end
+        component_index = len(models)
+    if component_index:
+        resume_after = models[component_index - 1]
+    else:
+        resume_after = None
+
+    if config.setting("multiprocess", False):
+        raise NotImplementedError(
+            "multiprocess component benchmarking is not yet implemented"
+        )
+    else:
+        open_pipeline(resume_after, mode="r")
+
+    for k in preload_injectables:
+        if inject.get_injectable(k, None) is not None:
+            logger.info("pre-loaded %s", k)
+
+    # Directories Logging
+    for k in ["configs_dir", "settings_file_name", "data_dir", "output_dir"]:
+        logger.info(f"DIRECTORY {k}: {inject.get_injectable(k, None)}")
+
+    # Settings Logging
+    log_settings = [
+        "checkpoints",
+        "chunk_training_mode",
+        "chunk_size",
+        "chunk_method",
+        "trace_hh_id",
+        "households_sample_size",
+        "check_for_variability",
+        "use_shadow_pricing",
+        "want_dest_choice_sample_tables",
+        "log_alt_losers",
+    ]
+    for k in log_settings:
+        logger.info(f"SETTING {k}: {config.setting(k)}")
+
+    logger.info("setup_component completed: %s", component_name)
+
+
+def run_component(component_name):
+    logger.info("run_component: %s", component_name)
+    try:
+        if config.setting("multiprocess", False):
+            raise NotImplementedError(
+                "multiprocess component benchmarking is not yet implemented"
```

Comment:
Looks like multiprocessing is implemented later in the file.  Having this error here and the commented out code below is confusing.  Removing the contradictory statements/code would be helpful.

Reply:
Sorry for the confusion.  I've added a clarifying comment in the code here for future developers (including future me).

Component level timings for multiprocess benchmarking are not generated using this code, which re-runs individual
components inside a timing loop.  Instead, those benchmarks are generated by running the model all the way through once during "setup", and then extracting the per-component timings from logs later.